### PR TITLE
Refine experimental Scopes API

### PR DESCRIPTION
**Pull request by trueadm**
This PR cleans up some of the types around the experimental Scopes API and simplies the data structure around how they're created. The main goal here is to cleanup the API for integration with `createEventHandle`.

With this PR, rather than create a `ScopeMethods` object, we create only a `ScopeInstance` object, which has the methods on – but attached from static functions rather than generated inline. In order for the methods to work with this change, it uses itself (the `ScopeInstance`) as a key to ask the renderer for it's associated fiber so it can carry out traversal.

### CODE REVIEW

File: packages/react-dom/src/client/ReactDOMComponentTree.js [hunk 1]
```diff
@@ -205,3 +206,9 @@ export function getEventListenerMap(node: EventTarget): ElementListenerMap {
   }
   return elementListenerMap;
 }
+
+export function getFiberFromScopeInstance(
+  scope: ReactScopeInstance,
+): null | Fiber {
+  return (scope: any)[internalInstanceKey] || null;
```
Comment (bvaughn) [inline packages/react-dom/src/client/ReactDOMComponentTree.js:213]: Wrap with feature flag?

**Comment by bvaughn** [comment]
What are the longer term plans for the Scope API and non-DOM renderers? Are there any?

**Comment by trueadm** [comment]
> What are the longer term plans for the Scope API and non-DOM renderers? Are there any?

There are no plans to use the Scope API in anything other than the DOM right now (they also work with the test renderer). We want to refine the API once we remove Flare, so that they move to become `useScope`, so around there we'll likely expose a larger design document as to how it might help in a larger picture with other renderers (if they are interested in the features).

**Comment by bvaughn** [comment]
> There are no plans to use the Scope API in anything other than the DOM right now

Interesting. I just saw all of the "noop" methods being added to host configs.

Wonder if we should create a `ReactFiberHostConfigWithNoScope` file that just exports shims for all of the Scopes APIs? That would be more inline with other things like hydration/mutation/persistence.

**Comment by trueadm** [comment]
@bvaughn Ah that's a good point. I can totally do that tomorrow :)

File: packages/react-reconciler/src/ReactFiberCommitWork.old.js [hunk 2]
```diff
@@ -1725,6 +1728,7 @@ function commitWork(current: Fiber | null, finishedWork: Fiber): void {
             updateDeprecatedEventListeners(nextListeners, finishedWork, null);
           }
         }
+        prepareScopeUpdate(scopeInstance, finishedWork);
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberCommitWork.old.js:1731]: Why is this necessary to have here? Shouldn't the call in complete work be sufficient?

As an aside, I *think* this would be the only place where we have an indirect call to `precacheFiberNode` in the commit phase.

Comment (trueadm) [reply packages/react-reconciler/src/ReactFiberCommitWork.old.js:1731]: This is necessary so we update the the fiber on scope. We do one in the complete phase for initial mount, and in the commit phase after each update (so we point to the current fiber tree). This is similar to the `updateFiberProps` logic we have in the other commit methods, except scopes don't have props.
